### PR TITLE
Fix endOf, daysInMonth, subtract and calendar persistence issues

### LINKED ISSUE #7: endOf and daysInMonth don't work properly
`endOf` and `daysInMonth` functions don't work when they are called inside the Esfand month of years with no extra leap day.

### PR DESCRIPTION
**Pull request by mpourismaiel**
UPDATE: Sorry my previous commit (removed by force push) didn't solve the problem. Also removed the PR note so this one is the new one.

- Fixed Esfand's `endOf` and `daysInMonth` functionality
- Fixed calendar persistence issue in some edge cases
- Fixed Farvardin's `subtract` issue
- Added tests

### Issues fixed

- fixes #7 

### Tests

```
 PASS  test/plugin.test.js
 PASS  test/relaiable-calendar-change.test.js
-------------|----------|----------|----------|----------|-------------------|
File         |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-------------|----------|----------|----------|----------|-------------------|
All files    |    98.15 |       96 |    97.06 |    98.13 |                   |
 calendar.js |    95.35 |    84.62 |      100 |    95.35 |           123,132 |
 constant.js |       90 |      100 |        0 |       90 |                17 |
 index.js    |      100 |      100 |      100 |      100 |                   |
 plugin.js   |    99.39 |    98.99 |      100 |    99.38 |               155 |
-------------|----------|----------|----------|----------|-------------------|
```

image

### CODE REVIEW

File: src/plugin.js [hunk 1]
```diff
@@ -139,13 +143,13 @@ export default (o, Dayjs, dayjs) => {
     }
     switch (unit) {
       case C.DATE:
-        instanceFactory(int, this.$jM)
+        innerSetDate(int, this.$jM)
         break
       case C.M:
-        instanceFactory(this.$jD, int)
+        innerSetDate(this.$jD, int === -1 ? 11 : int, int === -1 ? this.$jy - 1 : this.$jy)
```
Comment (farnabaz) [inline src/plugin.js]: I think it would be useful if you describe this line use case in a comment, so anyone else can understand why `int === -1` is so special.

Comment (mpourismaiel) [reply src/plugin.js]: Of course, will do.

**Comment by mpourismaiel** [comment]
@farnabaz Is there anything else wrong or problematic with the PR?

File: src/plugin.js (hunk 1 above)
Comment (mpourismaiel) [reply src/plugin.js]: It was also causing another error related to changes to month which would change the year. Wrote another test for it as well.

**Comment by farnabaz** [comment]
@mpourismaiel I add a test for `subtract` function

Also, I don't think this it is a good idea to manipulate `$set` function to handle subtract logic, there should be a simpler way to fix this issue

**Comment by mpourismaiel** [comment]
There is. We would need to change `calendar` module but that is out of the scope of my expertise. This is a dirty patch to fix the problem for now, a better fix would be to either make `calendar` module handle negative values.

**Comment by farnabaz** [comment]
Thanks 🍺

File: test/plugin.test.js [hunk 2]
```diff
@@ -262,14 +261,22 @@ describe('subtract', () => {
     expect(date2.date()).toEqual(date.date())
   })
 
+  test('add 11 months in the middle of the year', () => {
```
[the remark this answers is not in the document]
Comment (mpourismaiel) [reply test/plugin.test.js]: Sorry I forgot to press submit review on this one.

[the remark this answers is not in the document]
Comment (farnabaz) [reply test/plugin.test.js]: I do split tests into several files for better maintenance
